packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts

```diff
@@ -90,19 +134,19 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
      * Returns the GC data of this node. If nothing has changed since the last time we summarized, it tries to reuse
      * existing data.
      */
-    public async getGCData(): Promise<IGCData> {
+    public async getGCData(): Promise<IGarbageCollectionData> {
         assert(this.getGCDataFn !== undefined, "GC data cannot be retrieved without getGCDataFn");
 
-        if (!this.hasChanged()) {
-            // Nothing has changed since last summary. If we have the GC data from previous run, return it.
+        if (!super.hasChanged()) {
```

Comment:
It took me a minute to figure out why this makes sense, but maybe some clarification between the difference in `getGCData()`/`initialGCDetails` vs. `usedRoutes` would help.

I think `getGCData()` and `initialGCDetailsP` are referring to outbound dependencies/routes. i.e. the node is giving a list of all the GC nodes/routes it depends on.

But the `usedRoutes` and `referenceUsedRoutes` are referring to inbound dependencies/routes. i.e. all GC nodes of this node's subtree which are depended on by the root (or just "used"- should not be collected).

So this check is valid using `super.hasChanged` because we don't care about how we are referenced externally when explaining the GC nodes we depend on. Is that correct? Is it correct even if we can return dependencies between GC nodes within this subtree?

Reply:
I see how this is confusing. What you said is correct except for the `initialGCDetailsP` part. I think the name is to be blamed here which I will change :-). `initialGCDetailsP` is just providing the GCData and the usedRoutes from the summary before this client was created. Maybe this should be named `gcDetailsInInitialSummary`?

Another confusion is because of the name `hasChanged`. There are 2 things that can change in a node - The actual data in the node or the used state of a node which is basically the references to this node from outside. So, I think we need two separate methods - `hasReferenceChanged` or `hasUsedStateChanged` and `hasDataChanged`. `hasChanged` is true if either of these changed.

Then `getGCData` checks for `hasDataChanged` since it does not care if the external references to this node changed to generate the GC data.
`summarize` would check `hasChanged` because it needs to re-summarize even if only the external references to it changed.